lib/logflare/logs/search_operations.ex

```diff
@@ -80,24 +80,77 @@ defmodule Logflare.Logs.SearchOperations do
     %{so | query: query}
   end
 
-  def unnest_log_level(%_{source: %{id: source_id}} = so) do
-    source_schema = SourceSchemas.Cache.get_source_schema_by(source_id: source_id)
+  @doc """
+  Unnests and selects:
+
+  * recommended query fields (`bigquery_clustering_fields`, `suggested_keys`)
+  * `metadata.level`
+  """
+  @spec unnest_recommended_fields(SO.t()) :: SO.t()
+  def unnest_recommended_fields(%SO{source: source} = so) do
+    source_schema = SourceSchemas.Cache.get_source_schema_by(source_id: source.id)
     flatmap = Map.get(source_schema || %{}, :schema_flat_map)
 
-    if is_map_key(flatmap || %{}, "metadata.level") do
+    recommended_fields =
+      (Logflare.Sources.Source.recommended_query_fields(source) ++ ["metadata.level"])
+      |> Enum.uniq()
+
+    if Enum.empty?(recommended_fields) or is_nil(flatmap) do
+      so
+    else
+      {metadata_fields, top_level_fields} =
+        recommended_fields
+        |> Enum.filter(&Map.has_key?(flatmap, &1))
+        |> Enum.split_with(&String.starts_with?(&1, "metadata."))
+
+      metadata_field_atoms =
+        metadata_fields
+        |> Enum.map(fn "metadata." <> field -> String.to_atom(field) end)
+
+      top_level_field_atoms =
+        top_level_fields
+        |> Enum.map(&String.to_atom/1)
+
       query =
         so.query
-        |> join(:inner, [t], m in fragment("UNNEST(?)", t.metadata), on: true)
-        |> select_merge([..., m], %{
-          level: m.level
-        })
+        |> maybe_unnest_metadata(metadata_field_atoms)
+        |> maybe_select_top_level_fields(top_level_field_atoms)
 
       %{so | query: query}
-    else
-      so
     end
   end
 
+  defp maybe_unnest_metadata(query, []), do: query
+
+  defp maybe_unnest_metadata(query, metadata_field_atoms) do
```

Comment:
have we considered adding to the LQL string using the select rules like `s:metadata.level s:my_field s:my.nested_field` ? it could greatly simplify this as it looks like we're duplicating the LQL select rules logic.

Reply:
Hadn't considered this. Yes, it would be simpler. 

They won't be unnested, so are you thinking we add support to Lql for that?
Or we could leave them as flattened keys in the search results and handle it when building the links.